receiver/kafkareceiver/avro_unmarshaler.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	return a.mapLog(avroLog), nil
+}
+
+func (a *avroLogsUnmarshaler) mapLog(avroLog map[string]interface{}) plog.Logs {
+	logs := plog.NewLogs()
+	resourceLogs := logs.ResourceLogs().AppendEmpty()
+	resourceAttributes := resourceLogs.Resource().Attributes()
+	scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+	logRecord := scopeLogs.LogRecords().AppendEmpty()
+	logRecordAttributes := logRecord.Attributes()
+
+	for sourceField, target := range a.mapping {
+		switch {
+		case isResourceAttribute(target):
+			targetField := strings.TrimPrefix(target, resourceAttributesPrefix)
+			mapAttribute(avroLog, sourceField, &resourceAttributes, targetField)
+		case isLogRecordAttribute(target):
+			targetField := strings.TrimPrefix(target, logRecordAttributesPrefix)
+			mapAttribute(avroLog, sourceField, &logRecordAttributes, targetField)
+		case target == bodyField:
+			logRecord.Body().SetStr(avroLog[sourceField].(string))
+		case target == timestampField:
+			timestamp := avroLog[timestampField].(time.Time)
+			logRecord.SetTimestamp(pcommon.NewTimestampFromTime(timestamp))
+		case target == severityTextField:
+			logRecord.SetSeverityText(avroLog[sourceField].(string))
+		case target == severityNumberField:
+			severityNumber, ok := avroLog[sourceField].(int32)
+			if !ok {
+				severityNumber = 0
+			}
+			logRecord.SetSeverityNumber(plog.SeverityNumber(severityNumber))
+		}
+	}
+
+	return logs
+}
+
+func isResourceAttribute(targetField string) bool {
+	return strings.HasPrefix(targetField, resourceAttributesPrefix)
+}
+
+func isLogRecordAttribute(targetField string) bool {
+	return strings.HasPrefix(targetField, logRecordAttributesPrefix)
+}
+
+func mapAttribute(avroLog map[string]interface{}, sourceField string, attributes *pcommon.Map, targetField string) {
+	sourceValue := avroLog[sourceField]
+	if sourceValue != nil {
+		setAttribute(attributes, targetField, sourceValue)
+	}
+}
+
+func setAttribute(attributes *pcommon.Map, key string, value interface{}) {
+	attribute := attributes.PutEmpty(key)
+	_ = attribute.FromRaw(value)
+}
+
+func (a *avroLogsUnmarshaler) Init(schemaURL string, mapping map[string]string) error {
+	a.mapping = mapping
+
+	var schemaDeserializer avroDeserializer
+	var err error
+
+	if strings.HasPrefix(schemaURL, avroSchemaPrefixFile) {
+		schemaFilePath := strings.TrimPrefix(schemaURL, avroSchemaPrefixFile)
+		schemaFilePath = filepath.Clean(schemaFilePath)
+		schemaDeserializer, err = newAVROFileSchemaDeserializer(schemaFilePath)
+	} else {
+		// TODO: schema registry deserializer
+		err = fmt.Errorf("unimplemented schema url prefix")
+	}
+
+	a.deserializer = schemaDeserializer
+
+	return err
+}
+
+type avroDeserializer interface {
+	Deserialize([]byte) (map[string]interface{}, error)
+}
+
+type avroFileSchemaDeserializer struct {
+	codec *goavro.Codec
+}
+
+func newAVROFileSchemaDeserializer(path string) (avroDeserializer, error) {
+	schema, err := loadAVROSchemaFromFile(path)
```

Comment:
If possible, could we avoid having a strict method to read from disk when ideally we'd want to read the definition from the configuration?

If you want to make it easier to test with, I changing the method handler to:
```go
func readAVROSchema(r io.Reader) (schema, error)
```

This would allow you to use `bytes.NewBuffer`, `strings.Reader`, or `os.File` without needing to change the code path that does the main work and make it easier to test.

Reply:
Thanks for your comments, i'll address them.